hazelcast/src/test/java/com/hazelcast/client/listeners/AbstractListenersOnReconnectTest.java

```diff
@@ -185,61 +171,43 @@ private void testListenersTemporaryNetworkBlockage() {
-    public void testClusterReconnectDueToHeartbeatSmartRouting() {
-        factory.newHazelcastInstance();
-
-        ClientConfig clientConfig = getSmartClientConfigWithHeartbeat();
-        client = factory.newHazelcastClient(clientConfig);
-
-        testListenersHeartbeatTimeoutToCluster();
+    public void testListenersHeartbeatTimeoutToCluster_smart_singleServer() {
+        testListenersHeartbeatTimeoutToCluster(true, 1);
     }
 
     @Test
-    public void testClusterReconnectMultipleServersDueToHeartbeatSmartRouting() {
-        factory.newInstances(null, 3);
-
-        ClientConfig clientConfig = getSmartClientConfigWithHeartbeat();
-        client = factory.newHazelcastClient(clientConfig);
-
-        testListenersHeartbeatTimeoutToCluster();
+    public void testListenersHeartbeatTimeoutToCluster_nonSmart_singleServer() {
+        testListenersHeartbeatTimeoutToCluster(false, 1);
     }
 
     @Test
-    public void testClusterReconnectDueToHeartbeatNonSmartRouting() {
-        factory.newHazelcastInstance();
-
-        ClientConfig clientConfig = getNonSmartClientConfigWithHeartbeat();
-        client = factory.newHazelcastClient(clientConfig);
-
-        testListenersHeartbeatTimeoutToCluster();
+    public void testListenersHeartbeatTimeoutToCluster_smart_multipleServer() {
+        testListenersHeartbeatTimeoutToCluster(true, 3);
     }
 
     @Test
-    public void testClusterReconnectMultipleServerDueToHeartbeatNonSmartRouting() {
-        factory.newInstances(null, 3);
-
-        ClientConfig clientConfig = getNonSmartClientConfigWithHeartbeat();
-        client = factory.newHazelcastClient(clientConfig);
-
-        testListenersHeartbeatTimeoutToCluster();
+    public void testListenersHeartbeatTimeoutToCluster_nonSmart_multipleServer() {
+        testListenersHeartbeatTimeoutToCluster(false, 3);
     }
 
-    private void testListenersHeartbeatTimeoutToCluster() {
-        setupListener();
-
-        final CountDownLatch disconnectedLatch = new CountDownLatch(1);
-        final CountDownLatch connectedLatch = new CountDownLatch(1);
-        client.getLifecycleService().addLifecycleListener(new LifecycleListener() {
-            @Override
-            public void stateChanged(LifecycleEvent event) {
-                if (LifecycleEvent.LifecycleState.CLIENT_DISCONNECTED == event.getState()) {
-                    disconnectedLatch.countDown();
-                }
-                if (LifecycleEvent.LifecycleState.CLIENT_CONNECTED == event.getState()) {
-                    connectedLatch.countDown();
-                }
+    private void testListenersHeartbeatTimeoutToCluster(boolean isSmartClient, int nodeCount) {
+        factory.newInstances(null, nodeCount);
+        ClientConfig clientConfig = createClientConfig(isSmartClient);
+        ListenerConfig listenerConfig = new ListenerConfig();
+        CountDownLatch disconnectedLatch = new CountDownLatch(1);
+        CountDownLatch connectedLatch = new CountDownLatch(2);
```

Comment:
why not wait for the first connected event before blocking the network?

Reply:
We don't need to. When `factory.newHazelcastClient(clientConfig);` we are sure that client is connected and ready for the test. When the method returns client is connected and even got first membership event. 
We are counting both connected events because the firing connected event is offloaded to a different thread. 
In some very unlucky scenario, `CLIENT_CONNECTED` event could arrive late, but that does not mean client is not `CONNECTED`

Reply:
@ihsandemir Updated the pr to wait for CLIENT_CONNECTED event at the beginning.